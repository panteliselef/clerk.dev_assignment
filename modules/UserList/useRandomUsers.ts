import { useInfiniteQuery } from '@tanstack/react-query';
import { fetchRandomUsers, RandomUser } from '@services/api/random-user';

export const useRandomUsers = () =>
    useInfiniteQuery(
        ['users'],
        async ({ pageParam = 1 }) => {
            const data = await fetchRandomUsers(pageParam, 10);
            if (!data) {
                throw new Error('Users not found');
            }

            return {
                nextPage: pageParam + 1,
                data: data.results as RandomUser[],
            };
        },
        {
            getNextPageParam: ({ nextPage }) => (nextPage <= 10 ? nextPage : undefined),
        },
    );
