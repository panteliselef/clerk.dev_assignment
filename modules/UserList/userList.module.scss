@use 'styles/design/base/breakpoints' as brk;

.user_list {
    &_cont {
        display: flex;
        justify-content: flex-start;
        flex-direction: row;
        flex-wrap: nowrap;
        align-items: center;
        max-width: 1000px;
        padding: 0 1rem;
        width: 100%;
        gap: 1rem;

        scroll-snap-type: x mandatory;

        overflow: auto;

        @include brk.tablet2 {
            flex-direction: row;
            align-items: unset;
        }
    }

    &_linear {
        position: absolute;
        top: 0;
        right: 0;
        width: 10px;
        height: 100%;
        //background: linear-gradient(to right, rgba(255, 255, 255, 0.1), var(--bg-color));
    }
}
