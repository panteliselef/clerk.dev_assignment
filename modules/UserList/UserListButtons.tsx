import { useCarouselKeyboardEvents } from '@hooks/useCarousel';
import { useRandomUsers } from '@modules/UserList/useRandomUsers';
import Stack from '@layouts/Stack';
import { useDebouncedWidth } from '@hooks/useWindowDimensions';
import breakpoints from '@utils/breakpoints';
import ArrowLeft from '@components/icons/ArrowLeft';
import ArrowRight from '@components/icons/ArrowRight';

const UserListButtons = () => {
    const { scrollPrev, scrollNext } = useCarouselKeyboardEvents();
    const { isError } = useRandomUsers();

    return (
        <Stack
            direction={'row'}
            style={{
                width: '100%',
                padding: '0 1rem',
            }}
            alignItems={'center'}
            justifyContent={'space-between'}
        >
            <button disabled={isError} onClick={scrollPrev}>
                <ArrowLeft
                    style={{
                        fill: 'var(--text-color)',
                        width: '20px',
                    }}
                />
                Scroll Left
            </button>
            <button disabled={isError} onClick={scrollNext}>
                Scroll Right
                <ArrowRight
                    style={{
                        fill: 'var(--text-color)',
                        width: '20px',
                    }}
                />
            </button>
        </Stack>
    );
};

export default UserListButtons;

export const UserListDesktopNavigation = () => {
    const w = useDebouncedWidth();
    if (w < breakpoints.tablet2) return null;
    return <UserListButtons />;
};
