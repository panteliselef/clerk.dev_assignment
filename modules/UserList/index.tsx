import UserCard from '@components/UserCard';
import styles from './userList.module.scss';
import Stack from '@layouts/Stack';
import UserCardSkeleton from '@components/UserCard/skeleton';
import { useInView } from 'react-intersection-observer';
import { useCarousel, useCarouselKeyboardEvents } from '@hooks/useCarousel';
import { useDebouncedWidth } from '@hooks/useWindowDimensions';
import breakpoints from '@utils/breakpoints';
import { useRandomUsers } from '@modules/UserList/useRandomUsers';

const UserListError = () => (
    <p
        style={{
            border: '1px solid #C5283D',
            borderRadius: '4px',
            background: '#C5283D88',
            height: '100%',
            width: '100%',
            textAlign: 'center',
            padding: '2rem',
        }}
    >
        Users are unavailable
    </p>
);

const UserList = () => {
    const { ref: carouselRef } = useCarousel();
    const { data, isFetching, hasNextPage, isFetchingNextPage, fetchNextPage, isError } = useRandomUsers();

    const { ref } = useInView({
        /* Optional options */
        threshold: 0,
        triggerOnce: false,
        async onChange(inView) {
            if (inView && hasNextPage && !isFetchingNextPage) {
                await fetchNextPage();
            }
        },
    });

    const users = data?.pages.map(({ data }) => data).flat();

    return (
        <div ref={carouselRef} className={styles.user_list_cont}>
            {isError && <UserListError />}
            {users && users.map((user) => <UserCard key={user.name.first + user.name.last} {...user} />)}
            {isFetching && (
                <>
                    <UserCardSkeleton />
                    <UserCardSkeleton />
                    <UserCardSkeleton />
                </>
            )}
            {hasNextPage && !isFetching && <UserCardSkeleton ref={ref} />}
        </div>
    );
};

export const UserListButtons = () => {
    const { scrollPrev, scrollNext } = useCarouselKeyboardEvents();
    const { isError } = useRandomUsers();
    const w = useDebouncedWidth();

    if (w < breakpoints.tablet2) return null;
    return (
        <Stack
            direction={'row'}
            style={{
                width: '100%',
            }}
            alignItems={'center'}
            justifyContent={'space-between'}
        >
            <button data-testid="custom-element" disabled={isError} onClick={scrollPrev}>
                Prev
            </button>
            <button disabled={isError} onClick={scrollNext}>
                Next
            </button>
        </Stack>
    );
};

export default UserList;
