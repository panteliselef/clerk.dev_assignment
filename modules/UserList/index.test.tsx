import React from 'react';
import {act, render, screen, waitFor} from '@testing-library/react';
import UserList from '@modules/UserList/index';
import {useRandomUsers} from '@modules/UserList/useRandomUsers';

import renderer from 'react-test-renderer';
import UserListButtons, {UserListDesktopNavigation} from "@modules/UserList/UserListButtons";
// Make TypeScript Happy, by resolving TS errors
const mockedUseUsersQuery = useRandomUsers as jest.Mock<any>;

// Mock the hook module
jest.mock('@modules/UserList/useRandomUsers');

const mockUsers = [
    {
        data: [
            {
                name: {title: 'Miss', first: 'Laura', last: 'Woods'},
                location: {
                    street: {number: 2479, name: 'Henry Street'},
                    city: 'Blessington',
                    state: 'Wexford',
                    country: 'Ireland',
                    postcode: 78276,
                    coordinates: {latitude: '2.0565', longitude: '95.2422'},
                    timezone: {offset: '+1:00', description: 'Brussels, Copenhagen, Madrid, Paris'},
                },
                email: 'laura.woods@example.com',
                phone: '031-623-5189',
                id: {name: 'PPS', value: '1101776T'},
                picture: {
                    large: 'https://randomuser.me/api/portraits/women/88.jpg',
                    medium: 'https://randomuser.me/api/portraits/med/women/88.jpg',
                    thumbnail: 'https://randomuser.me/api/portraits/thumb/women/88.jpg',
                },
            },
            {
                name: {title: 'Mr', first: 'Marten', last: 'Faber'},
                location: {
                    street: {number: 6167, name: 'Grüner Weg'},
                    city: 'Falkenberg/Elster',
                    state: 'Thüringen',
                    country: 'Germany',
                    postcode: 99553,
                    coordinates: {latitude: '89.4367', longitude: '135.6354'},
                    timezone: {offset: '+5:45', description: 'Kathmandu'},
                },
                email: 'marten.faber@example.com',
                phone: '0100-8354415',
                id: {name: 'SVNR', value: '18 010860 F 495'},
                picture: {
                    large: 'https://randomuser.me/api/portraits/men/1.jpg',
                    medium: 'https://randomuser.me/api/portraits/med/men/1.jpg',
                    thumbnail: 'https://randomuser.me/api/portraits/thumb/men/1.jpg',
                },
            },
        ],
    },
];

describe('<UserListButtons />', () => {
    beforeEach(() => {
        mockedUseUsersQuery.mockImplementation(() => ({isLoading: true}));
    });
    afterEach(() => {
        jest.clearAllMocks();
    });

    it('Renders without crashing', () => {
        render(<UserListButtons/>);
    });

    it('Disabled buttons if error with fetching', async () => {
        mockedUseUsersQuery.mockImplementation(() => ({
            isError: true,
        }));

        render(<UserListButtons/>);
        const buttonLeft = screen.getByText(/Left/i);
        expect(buttonLeft).toHaveAttribute('disabled');

        const buttonRight = screen.getByText(/Right/i);
        expect(buttonRight).toHaveAttribute('disabled');
    });
});

describe('<UserListDesktopNavigation />', () => {

    it('Renders without crashing', () => {
        render(<UserListDesktopNavigation/>);
    });

    it('Hidden buttons on mobile', async () => {
        render(<UserListDesktopNavigation/>);

        await act(() => {
            global.window.innerWidth = 300;
            global.window.dispatchEvent(new Event('resize'));
        });

        await waitFor(() => expect(screen.queryByRole('button')).not.toBeInTheDocument());
    });
});

describe('<UserList />', () => {
    beforeEach(() => {
        mockedUseUsersQuery.mockImplementation(() => ({isFetching: true}));
    });
    afterEach(() => {
        jest.clearAllMocks();
    });

    it('Renders without crashing', () => {
        render(<UserList/>);
    });

    it('Display 3 skeletons on Loading', async () => {
        const tree = renderer.create(<UserList/>).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('Display Error if fetching failed', async () => {
        mockedUseUsersQuery.mockImplementation(() => ({isError: true}));
        const tree = renderer.create(<UserList/>).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('Display 2 user cards', () => {
        mockedUseUsersQuery.mockImplementation(() => ({
            data: {
                pages: mockUsers,
            },
        }));
        const tree = renderer.create(<UserList/>).toJSON();
        expect(tree).toMatchSnapshot();
    });
});
