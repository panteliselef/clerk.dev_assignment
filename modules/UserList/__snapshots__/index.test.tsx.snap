// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UserList /> Display 2 user cards 1`] = `
<div
  className="user_list_cont"
>
  <div
    className="user_card_cont v-flex align-center"
  >
    <div
      style={
        {
          "position": "relative",
        }
      }
    >
      <img
        alt="profile pic for user Laura"
        className="user_card_img"
        data-nimg="1"
        decoding="async"
        height={60}
        loading="lazy"
        onError={[Function]}
        onLoad={[Function]}
        src="/_next/image?url=https%3A%2F%2Frandomuser.me%2Fapi%2Fportraits%2Fwomen%2F88.jpg&w=128&q=75"
        srcSet="/_next/image?url=https%3A%2F%2Frandomuser.me%2Fapi%2Fportraits%2Fwomen%2F88.jpg&w=64&q=75 1x, /_next/image?url=https%3A%2F%2Frandomuser.me%2Fapi%2Fportraits%2Fwomen%2F88.jpg&w=128&q=75 2x"
        style={
          {
            "color": "transparent",
          }
        }
        width={60}
      />
      <span
        aria-label="Miss"
        className="user_card_pill"
      >
        Miss
      </span>
    </div>
    <div
      className=" asCol"
      style={
        {
          "alignItems": undefined,
          "gap": 3,
          "justifyContent": undefined,
          "width": "100%",
        }
      }
    >
      <p
        aria-label="Laura Woods"
        className="user_card_name"
      >
        Laura
         
        Woods
      </p>
      <a
        className="user_card_location"
        href="https://maps.google.com/?q=2479 Henry Street, Blessington, Wexford, 78276"
        rel="noopener noreferrer"
        tabIndex={1}
        target="_blank"
      >
        Ireland, Ireland
      </a>
    </div>
    <div
      className=" asRow as_row_tablet_col"
      style={
        {
          "alignItems": undefined,
          "gap": 10,
          "justifyContent": undefined,
          "width": "100%",
        }
      }
    >
      <a
        className="user_card_button"
        href="mailto://laura.woods@example.com"
        rel="noopener noreferrer"
        style={
          {
            "flex": "1 1 100px",
          }
        }
        tabIndex={1}
        target="_blank"
      >
        Email
      </a>
      <a
        className="user_card_button"
        href="tel://031-623-5189"
        rel="noopener noreferrer"
        tabIndex={1}
        target="_blank"
      >
        <svg
          className="w-6 h-6"
          fill="none"
          stroke="var(--text-color)"
          strokeWidth="1.5"
          viewBox="0 0 24 24"
          width={20}
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M20.25 3.75v4.5m0-4.5h-4.5m4.5 0l-6 6m3 12c-8.284 0-15-6.716-15-15V4.5A2.25 2.25 0 014.5 2.25h1.372c.516 0 .966.351 1.091.852l1.106 4.423c.11.44-.054.902-.417 1.173l-1.293.97a1.062 1.062 0 00-.38 1.21 12.035 12.035 0 007.143 7.143c.441.162.928-.004 1.21-.38l.97-1.293a1.125 1.125 0 011.173-.417l4.423 1.106c.5.125.852.575.852 1.091V19.5a2.25 2.25 0 01-2.25 2.25h-2.25z"
            strokeLinecap="round"
            strokeLinejoin="round"
          />
        </svg>
      </a>
    </div>
  </div>
  <div
    className="user_card_cont v-flex align-center"
  >
    <div
      style={
        {
          "position": "relative",
        }
      }
    >
      <img
        alt="profile pic for user Marten"
        className="user_card_img"
        data-nimg="1"
        decoding="async"
        height={60}
        loading="lazy"
        onError={[Function]}
        onLoad={[Function]}
        src="/_next/image?url=https%3A%2F%2Frandomuser.me%2Fapi%2Fportraits%2Fmen%2F1.jpg&w=128&q=75"
        srcSet="/_next/image?url=https%3A%2F%2Frandomuser.me%2Fapi%2Fportraits%2Fmen%2F1.jpg&w=64&q=75 1x, /_next/image?url=https%3A%2F%2Frandomuser.me%2Fapi%2Fportraits%2Fmen%2F1.jpg&w=128&q=75 2x"
        style={
          {
            "color": "transparent",
          }
        }
        width={60}
      />
      <span
        aria-label="Mr"
        className="user_card_pill"
      >
        Mr
      </span>
    </div>
    <div
      className=" asCol"
      style={
        {
          "alignItems": undefined,
          "gap": 3,
          "justifyContent": undefined,
          "width": "100%",
        }
      }
    >
      <p
        aria-label="Marten Faber"
        className="user_card_name"
      >
        Marten
         
        Faber
      </p>
      <a
        className="user_card_location"
        href="https://maps.google.com/?q=6167 Grüner Weg, Falkenberg/Elster, Thüringen, 99553"
        rel="noopener noreferrer"
        tabIndex={1}
        target="_blank"
      >
        Germany, Germany
      </a>
    </div>
    <div
      className=" asRow as_row_tablet_col"
      style={
        {
          "alignItems": undefined,
          "gap": 10,
          "justifyContent": undefined,
          "width": "100%",
        }
      }
    >
      <a
        className="user_card_button"
        href="mailto://marten.faber@example.com"
        rel="noopener noreferrer"
        style={
          {
            "flex": "1 1 100px",
          }
        }
        tabIndex={1}
        target="_blank"
      >
        Email
      </a>
      <a
        className="user_card_button"
        href="tel://0100-8354415"
        rel="noopener noreferrer"
        tabIndex={1}
        target="_blank"
      >
        <svg
          className="w-6 h-6"
          fill="none"
          stroke="var(--text-color)"
          strokeWidth="1.5"
          viewBox="0 0 24 24"
          width={20}
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M20.25 3.75v4.5m0-4.5h-4.5m4.5 0l-6 6m3 12c-8.284 0-15-6.716-15-15V4.5A2.25 2.25 0 014.5 2.25h1.372c.516 0 .966.351 1.091.852l1.106 4.423c.11.44-.054.902-.417 1.173l-1.293.97a1.062 1.062 0 00-.38 1.21 12.035 12.035 0 007.143 7.143c.441.162.928-.004 1.21-.38l.97-1.293a1.125 1.125 0 011.173-.417l4.423 1.106c.5.125.852.575.852 1.091V19.5a2.25 2.25 0 01-2.25 2.25h-2.25z"
            strokeLinecap="round"
            strokeLinejoin="round"
          />
        </svg>
      </a>
    </div>
  </div>
</div>
`;

exports[`<UserList /> Display 3 skeletons on Loading 1`] = `
<div
  className="user_list_cont"
>
  <div
    className="user_card_cont v-flex align-center"
  >
    <div
      className="user_card_img user_card__skeleton_item"
    />
    <div
      className="user_card_name user_card__skeleton_item"
    />
    <div
      className="user_card_location user_card__skeleton_item"
    />
    <div
      className=" asRow as_row_tablet_col"
      style={
        {
          "alignItems": undefined,
          "gap": "8px",
          "justifyContent": undefined,
          "width": "100%",
        }
      }
    >
      <div
        className="user_card_button user_card__skeleton_item"
      />
      <div
        className="user_card_button user_card__skeleton_item"
      />
    </div>
  </div>
  <div
    className="user_card_cont v-flex align-center"
  >
    <div
      className="user_card_img user_card__skeleton_item"
    />
    <div
      className="user_card_name user_card__skeleton_item"
    />
    <div
      className="user_card_location user_card__skeleton_item"
    />
    <div
      className=" asRow as_row_tablet_col"
      style={
        {
          "alignItems": undefined,
          "gap": "8px",
          "justifyContent": undefined,
          "width": "100%",
        }
      }
    >
      <div
        className="user_card_button user_card__skeleton_item"
      />
      <div
        className="user_card_button user_card__skeleton_item"
      />
    </div>
  </div>
  <div
    className="user_card_cont v-flex align-center"
  >
    <div
      className="user_card_img user_card__skeleton_item"
    />
    <div
      className="user_card_name user_card__skeleton_item"
    />
    <div
      className="user_card_location user_card__skeleton_item"
    />
    <div
      className=" asRow as_row_tablet_col"
      style={
        {
          "alignItems": undefined,
          "gap": "8px",
          "justifyContent": undefined,
          "width": "100%",
        }
      }
    >
      <div
        className="user_card_button user_card__skeleton_item"
      />
      <div
        className="user_card_button user_card__skeleton_item"
      />
    </div>
  </div>
</div>
`;

exports[`<UserList /> Display Error if fetching failed 1`] = `
<div
  className="user_list_cont"
>
  <p
    style={
      {
        "background": "#C5283D88",
        "border": "1px solid #C5283D",
        "borderRadius": "4px",
        "height": "100%",
        "padding": "2rem",
        "textAlign": "center",
        "width": "100%",
      }
    }
  >
    Users are unavailable
  </p>
</div>
`;
